### PR TITLE
hdfs: support kerberos authentication #42

### PR DESCRIPTION
**Pull request by urykhy**
#### What is the purpose of this change?

access HDFS cluster with KERBEROS authentication

#### Was the change discussed in an issue or in the forum before?

[yes](https://github.com/rclone/rclone/issues/42#issuecomment-759630846)

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/hdfs/fs.go [hunk 1]
```diff
@@ -35,11 +83,26 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
...
 
-	client, err := hdfs.NewClient(hdfs.ClientOptions{
+	options := hdfs.ClientOptions{
 		Addresses:           []string{opt.Namenode},
-		User:                opt.Username,
 		UseDatanodeHostname: false,
-	})
+	}
+
+	if len(opt.ServicePrincipalName) > 0 {
```
Comment (ncw) [inline backend/hdfs/fs.go:91]: `if opt.ServicePrincipalName != "" {` would probably be more idiomatic. It compiles to the same code anyway!

File: backend/hdfs/fs.go [hunk 2]
```diff
@@ -35,11 +83,26 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
...
+		options.KerberosClient, err = getKerberosClient()
+		if err != nil {
+			return nil, fmt.Errorf("Problem with kerberos authentication: %s", err)
+		}
+		options.KerberosServicePrincipleName = opt.ServicePrincipalName
+
+		if len(opt.DataTransferProtection) > 0 {
```
Comment (ncw) [inline backend/hdfs/fs.go:98]: ...same here

File: backend/hdfs/hdfs.go [hunk 3]
```diff
@@ -32,6 +32,25 @@ func init() {
 				Value: "root",
 				Help:  "Connect to hdfs as root",
 			}},
+		}, {
+			Name:     "service_principal_name",
+			Help:     "Enables KERBEROS authentication. Specifies the Service Principle Name (<SERVICE>/<FQDN>) for the namenode.",
```
Comment (ncw) [inline backend/hdfs/hdfs.go:37]: Note that we try to wrap the help at 80 characters. The first line becomes the command line help for the flag.

You can use the backquote character for writing strings over multiple lines, so I'd write


```go
Help: `Kerberos service principal name for the namenode.

Enables KERBEROS authentication. Specifies the Service Principal Name
(<SERVICE>/<FQDN>) for the namenode.
```

File: backend/hdfs/hdfs.go [hunk 4]
```diff
@@ -32,6 +32,25 @@ func init() {
...
+			Required: false,
+			Examples: []fs.OptionExample{{
+				Value: "hdfs/namenode.hadoop.docker",
+				Help:  "Namenode running as service 'hdfs' with FQDN 'namenode.hadoop.docker'.",
+			}},
+			Advanced: true,
+		}, {
+			Name: "data_transfer_protection",
+			Help: "Specifies whether or not authentication, data signature integrity checks, and wire encryption is required when communicating the the datanodes." +
```
Comment (ncw) [inline backend/hdfs/hdfs.go:46]: Likewise here

```go
Help: `Kerberos data transfer protection: authentication|integrity|privacy

Specifies whether or not authentication, data signature integrity
checks, and wire encryption is required when communicating the the
datanodes. Possible values are 'authentication', 'integrity' and
'privacy'. Used only with KERBEROS enabled.
`
```

File: backend/hdfs/fs.go [hunk 5]
```diff
@@ -35,11 +83,26 @@ func NewFs(ctx context.Context, name, root string, m configmap.Mapper) (fs.Fs, e
...
 		UseDatanodeHostname: false,
-	})
+	}
+
+	if len(opt.ServicePrincipalName) > 0 {
+		options.KerberosClient, err = getKerberosClient()
+		if err != nil {
+			return nil, fmt.Errorf("Problem with kerberos authentication: %s", err)
+		}
+		options.KerberosServicePrincipleName = opt.ServicePrincipalName
```
Comment (ncw) [inline backend/hdfs/fs.go:96]: Principle or Principal? We should stick to one spelling!

I think `principal` is correct, which means that `hdfs.ClientOptions` is spelled wrong!

**Review by ncw — COMMENTED** [review]
I think this basically looks fine :-)

See inline for some minor spelling/layout nits.

Can we add this to the integration tests easily?

PS Forgot to say - thank you for doing this work - it is in demand by users :-)

File: backend/hdfs/fs.go (hunk 5 above)
Comment (urykhy) [reply backend/hdfs/fs.go:96]: yes.

should we replicate tipo from hdfs library ?

**Comment by urykhy** [comment]
> Can we add this to the integration tests easily?

i've add option to use kerberos in tests.

File: backend/hdfs/hdfs.go [hunk 6]
```diff
@@ -33,8 +33,11 @@ func init() {
 				Help:  "Connect to hdfs as root",
 			}},
 		}, {
-			Name:     "service_principal_name",
-			Help:     "Enables KERBEROS authentication. Specifies the Service Principle Name (<SERVICE>/<FQDN>) for the namenode.",
+			Name: "service_principal_name",
+			Help: `Kerberos service principal name for the namenode
+
+Enables KERBEROS authentication. Specifies the Service Principle Name
```
Comment (ncw) [inline backend/hdfs/hdfs.go:39]: Principal here!

File: backend/hdfs/fs.go (hunk 5 above)
Comment (ncw) [reply backend/hdfs/fs.go:96]: That is in the exported struct from the hdfs library isn't it? We'll have to live with that one!

File: docs/content/hdfs.md [hunk 7]
```diff
@@ -187,7 +187,10 @@ Here are the advanced options specific to hdfs (Hadoop distributed file system).
 
 #### --hdfs-service-principal-name
 
-Enables KERBEROS authentication. Specifies the Service Principle Name (<SERVICE>/<FQDN>) for the namenode.
+Kerberos service principal name for the namenode
+
+Enables KERBEROS authentication. Specifies the Service Principle Name
```
Comment (ncw) [inline docs/content/hdfs.md:192]: Principal here

**Review by ncw — COMMENTED** [review]
I spotted some more `Principles` inline!

Well done making a test - that looks great.

Do we need to test with and without Kerberos? If you think testing with Kerberos only is sufficient I'm happy to go with that.

File: docs/content/hdfs.md (hunk 7 above)
Comment (urykhy) [reply docs/content/hdfs.md:192]: yes.
`KerberosServicePrincipleName` comes from library.

**Comment by urykhy** [comment]
> Do we need to test with and without Kerberos? If you think testing with Kerberos only is sufficient I'm happy to go with that.

Kerberos tests is a bit more complicated to run (local krb5.conf required).
So... i'm in doubt.

**Comment by ncw** [comment]
> > Do we need to test with and without Kerberos? If you think testing with Kerberos only is sufficient I'm happy to go with that.
> 
> Kerberos tests is a bit more complicated to run (local krb5.conf required).
> So... i'm in doubt.

What I'll do is merge it as it is and we can see how the integration tests go.

I'll pull and push the updated image.

Thank you very much for doing this @urykhy

**Pull request merged**